src/unitful.jl

```diff
@@ -0,0 +1,293 @@
+    if aspect_ratio in (:auto, :none)
+        # Keep the default behavior (let Plots figure it out)
+        return
+    end
+    if aspect_ratio === :equal
+        aspect_ratio = 1
+    end
+    #=======================================================================================
+    Implementation example:
+
+    Consider an x axis in `u"m"` and a y axis in `u"s"`, and an `aspect_ratio` in `u"m/s"`.
+    On the first pass, `axisletter` is `:x`, so `aspect_ratio` is converted to `u"m/s"/u"m"
+    = u"s^-1"`. On the second pass, `axisletter` is `:y`, so `aspect_ratio` becomes
+    `u"s^-1"*u"s" = 1`. If at this point `aspect_ratio` is *not* unitless, an error has been
+    made, and the default aspect ratio fixing of Plots throws a `DimensionError` as it tries
+    to compare `0 < 1u"m"`.
+    =======================================================================================#
+    if axisletter === :y
+        attr[:aspect_ratio] = aspect_ratio * u
+        return
+    end
+    if axisletter === :x
+        attr[:aspect_ratio] = aspect_ratio / u
+        return
+    end
+    return
+end
+
+# Markers / lines
+function fixmarkercolor!(attr)
+    u = ustripattribute!(attr, :marker_z)
+    ustripattribute!(attr, :clims, u)
+    u == Unitful.NoUnits || append_unit_if_needed!(attr, :colorbar_title, u)
+end
+fixmarkersize!(attr) = ustripattribute!(attr, :markersize)
+fixlinecolor!(attr) = ustripattribute!(attr, :line_z)
+
+# strip unit from attribute[key]
+ustripattribute!(attr, key) =
+    if haskey(attr, key)
+        v = attr[key]
+        u = unit(eltype(v))
+        attr[key] = ustrip.(u, v)
+        return u
+    else
+        return Unitful.NoUnits
+    end
+# If supplied, use the unit (optional 3rd argument)
+function ustripattribute!(attr, key, u)
+    if haskey(attr, key)
+        v = attr[key]
+        if eltype(v) <: Quantity
+            attr[key] = ustrip.(u, v)
+        end
+    end
+    u
+end
+
+#=======================================
+Label string containing unit information
+=======================================#
+
+abstract type AbstractProtectedString <: AbstractString end
+struct ProtectedString{S} <: AbstractProtectedString
+    content::S
+end
+struct UnitfulString{S,U} <: AbstractProtectedString
+    content::S
+    unit::U
+end
+# Minimum required AbstractString interface to work with Plots
+const S = AbstractProtectedString
+Base.iterate(n::S) = iterate(n.content)
+Base.iterate(n::S, i::Integer) = iterate(n.content, i)
+Base.codeunit(n::S) = codeunit(n.content)
+Base.ncodeunits(n::S) = ncodeunits(n.content)
+Base.isvalid(n::S, i::Integer) = isvalid(n.content, i)
+Base.pointer(n::S) = pointer(n.content)
+Base.pointer(n::S, i::Integer) = pointer(n.content, i)
+"""
+    P_str(s)
+
+Creates a string that will be Protected from recipe passes.
+
+Example:
+```julia
+julia> plot(rand(10)*u"m", xlabel=P"This label is protected")
+
+julia> plot(rand(10)*u"m", xlabel=P"This label is not")
```

Comment:
I don't get this. In both cases `P` is being used. Also what is the effect of a label being protected ?
I guess it won't show units by default then?

Reply:
Fixed.